src/interpreter/ByteCodeInterpreter.cpp

```diff
@@ -2974,18 +2974,43 @@ NEVER_INLINE void InterpreterSlowPath::initializeGlobalVariable(ExecutionState&
 NEVER_INLINE void InterpreterSlowPath::createObjectOperation(ExecutionState& state, CreateObject* code, ByteCodeBlock* byteCodeBlock, Value* registerFile)
 {
     if (code->m_dataRegisterIndex != REGISTER_LIMIT) {
-        CreateObjectPrepare::CreateObjectData* data = reinterpret_cast<CreateObjectPrepare::CreateObjectData*>(registerFile[code->m_dataRegisterIndex].payload());
+        CreateObjectPrepare::CreateObjectData* data;
+        if (byteCodeBlock->codeBlock()->isAsyncOrGenerator()) {
+            data = reinterpret_cast<CreateObjectPrepare::CreateObjectData*>(registerFile[code->m_dataRegisterIndex].payload());
+        } else {
+            data = reinterpret_cast<CreateObjectPrepare::CreateObjectData*>(&registerFile[code->m_dataRegisterIndex]);
+        }
+
         Object* obj = registerFile[code->m_registerIndex].asObject();
-        ObjectStructureItemTightVector properties;
-        properties.reset(data->m_properties.takeBuffer(), data->m_values.size());
-        obj->m_structure = ObjectStructure::create(state.context(), std::move(properties));
+        if (!data->m_wasStructureComputed) {
+            size_t propertyCount = data->m_properties.size();
+            Optional<ObjectStructure*> cache;
+            if (data->m_allPrecomputed) {
+                cache = state.context()->vmInstance()->findRootedObjectStructure(data->m_properties.data(), propertyCount);
+            }
+            if (cache) {
+                obj->m_structure = cache.value();
+            } else {
+                obj->m_structure = ObjectStructure::create(state.context(),
+                                                           ObjectStructureItemTightVector(data->m_properties.data(), data->m_properties.data() + propertyCount),
+                                                           data->m_allPrecomputed);
+                if (data->m_canStoreStructureOnCode) {
+                    auto structure = obj->m_structure;
+                    data->m_initCode->m_cachedObjectStructure = structure;
+                    byteCodeBlock->m_otherLiteralData.pushBack(structure);
+                    state.context()->vmInstance()->addObjectStructureToRootSet(structure);
```

Comment:
I think that caching `structure` in the `byteCodeBlock->m_otherLiteralData` is unnecessary
because this structure will be cached and traced in the VMInstance too

Reply:
Good point!